### PR TITLE
Switch to in-tree rustc dependencies with a cfg flag

### PR DESCRIPTION
**Pull request by HKalbasi**
We can use this flag to detect and prevent breakages in rustc CI. (see #14846 and #15569)

~The `IN_RUSTC_REPOSITORY` is just a placeholder. Is there any existing cfg flag that rustc CI sets?~

### CODE REVIEW

**Comment by lnicola** [comment]
> Is there any existing cfg flag that rustc CI sets?

It enables the `in-rust-tree` feature, IIUC.

**Comment by HKalbasi** [comment]
This will probably make syncs harder, but I think it does worth it. Feel free to revert if it ended up breaking something or caused too much trouble.

@bors r+

**Comment by lnicola** [comment]
> This will probably make syncs harder, but I think it does worth it.

I hope not :grimacing:.

**Comment by Veykril** [comment]
Not a fan of the implications this has for our dependency graph. This is probably fine right now but won't scale in the future.

File: crates/parser/src/lib.rs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 #![warn(rust_2018_idioms, unused_lifetimes, semicolon_in_expressions_from_macros)]
 #![allow(rustdoc::private_intra_doc_links)]
+#![cfg_attr(feature = "in-rust-tree", feature(rustc_private))]
```
Comment (Veykril) [inline crates/parser/src/lib.rs:22]: why is this required?

File: crates/rust-analyzer/Cargo.toml [hunk 2]
```diff
@@ -89,4 +91,12 @@ mbe.workspace = true
...
-in-rust-tree = ["sysroot-abi", "ide/in-rust-tree", "syntax/in-rust-tree"]
+in-rust-tree = [
+    "sysroot-abi",
+    "ide/in-rust-tree",
+    "syntax/in-rust-tree",
+    "parser/in-rust-tree",
+    "rustc-dependencies/in-rust-tree",
+    "hir-def/in-rust-tree",
+    "hir-ty/in-rust-tree",
+]
```
Comment (Veykril) [inline crates/rust-analyzer/Cargo.toml:102]: Not a fan of this either, this will ultimately end up with mentioning almost all of our crates at some point

File: crates/hir-def/src/lib.rs [hunk 3]
```diff
@@ -8,6 +8,7 @@
 //! actually true.
 
 #![warn(rust_2018_idioms, unused_lifetimes, semicolon_in_expressions_from_macros)]
+#![cfg_attr(feature = "in-rust-tree", feature(rustc_private))]
```
Comment (Veykril) [inline crates/hir-def/src/lib.rs:11]: Why is this needed?

File: crates/syntax/src/lib.rs [hunk 4]
```diff
@@ -19,6 +19,7 @@
 //! [RFC]: <https://github.com/rust-lang/rfcs/pull/2256>
 //! [Swift]: <https://github.com/apple/swift/blob/13d593df6f359d0cb2fc81cfaac273297c539455/lib/Syntax/README.md>
 
+#![cfg_attr(feature = "in-rust-tree", feature(rustc_private))]
```
Comment (Veykril) [inline crates/syntax/src/lib.rs:22]: same here

**Review by Veykril — COMMENTED** [review]
While this is probably a solution, I have to say that I really don't like these changes 😅

File: crates/hir-def/src/lib.rs (hunk 3 above)
Comment (HKalbasi) [reply crates/hir-def/src/lib.rs:11]: This feature is needed even if you import the rustc private crates indirectly, which I think is unnecessarily too aggressive. Is there any way to depend on the crates in the rustc workspace without this feature? Clippy and Miri use this, but they are fine with nightly and feature flags.

File: crates/rust-analyzer/Cargo.toml (hunk 2 above)
Comment (HKalbasi) [reply crates/rust-analyzer/Cargo.toml:102]: Me too, maybe a non cargo cfg is a better fit here?

Comment (Veykril) [reply crates/rust-analyzer/Cargo.toml:102]: Yes, I think it'd be better if we could make it so that this doesn't thread through all the cargo files if possible

File: crates/hir-def/src/lib.rs (hunk 3 above)
Comment (Veykril) [reply crates/hir-def/src/lib.rs:11]: Ah right, since we might mention the items from those crates now. Well that's unfortunate but not much we can do in that case. Would be nice  to add a comment regarding that then.
